### ISSUE TITLE
tsv files

### ISSUE THREAD

**Issue by devloic** [mvtiaine/audacious-uade#34]
what is the difference between :
1. https://github.com/mvtiaine/audacious-uade/blob/master/conf/songdb/demozoo.tsv
and 
2. https://github.com/mvtiaine/audacious-uade/blob/master/tools/songdb/sources/demozoo.tsv

1 : 1MB , used by songdb binary
2 : 28MB, human readable

How can I convert 1 to 2 and 2 to 1 ?

**Comment by mvtiaine** [comment]
The files under https://github.com/mvtiaine/audacious-uade/tree/master/conf/songdb are generated with scala-cli script(s):
https://github.com/mvtiaine/audacious-uade/blob/master/tools/songdb/songdb.sc

It uses the tsv files under https://github.com/mvtiaine/audacious-uade/tree/master/tools/songdb/sources but note that it will also need some additional files not included in github, specifically local "mirror" of some of the AMP webpages. You can comment out the AMP related things of course.

The demozoo.tsv is generated with an SQL script (https://github.com/mvtiaine/audacious-uade/blob/master/tools/songdb/scripts/demozoo.sql) from demozoo postgres database dump (http://data.demozoo.org/demozoo-export.sql.gz)

All the other source tsv files are generated using the precalc binary+script https://github.com/mvtiaine/audacious-uade/blob/master/src/plugin/cli/precalc/precalc.sh from my local copy/mirror of the various sites/sources.

So to be specific, you convert from 2 to 1 with the Scala CLI scripts, but converting back is currently not possible (and not all fields are preserved either when going from 2 to 1).

**Comment by devloic** [comment]
Thank you for this detailed explanation. 

I am really impressed by your tenacity to "merge" so many different datasources that were not designed to be merged and with some of them going back to the 90s. I was about to try to do the same but when I saw all this legacy mess out there I felt this was going to take me another month of development and investigation. So I dug again into your code and with your indications I realize that audacious-uade includes a songdb binary that can give you metadata about almost any referenced amiga music file including song length which is huge.

I am working on extending azuracast ( a self-hosted web radio streamer) so it can stream amiga music with uade and I really wanted to show more than just the filename while the track gets streamed so now it is just a question of embedding songdb with liquidsoap. 

Besides that I'd like to contribute in developing one or two features that I think could be of use :

1. an option to export to sql or directly to sqlite so your data can be queried with other criterias then just the md5 but also: by author, by year , by format etc. 
2. an easy way to add entries to your database ( custom.tsv or using 1 for updating an sql database )

Now, one thing that could definitely nail the problem of having so many different music file formats is to define some kind of amiga (?) "container" format that would embed the existing music file along with its metadata. The author of uade tried such initiative ( https://gitlab.com/uade-music-player/rmc ) . Maybe with a player backed solution like audacious this would encourage adoption.

**Comment by mvtiaine** [comment]
I have on my TODO list for the Scala CLI tool to support alternative "merged" TSV format, which would be kind of intermediate between the "raw" TSVs and the "almost binary" conf TSVs. Those should then be easy to import to sqlite or parse with grep, python or whatever. Only thing is that some author parsing logic etc. still happens in the c++ side during loading, and that would need to be implemented in the Scala converter instead: https://github.com/mvtiaine/audacious-uade/tree/master/src/songdb/sources

Actually also the "compressed" songdb originally was a bit like that: https://github.com/mvtiaine/audacious-uade/blob/75a4b6192fe9f0822ed25e63ff8b38471908985c/conf/songdb/
before changing to the current format, as the songdb init was getting quite heavy on slower machines: https://github.com/mvtiaine/audacious-uade/pull/18

Alternative could also be to extend the songdb CLI binary to dump the whole in-memory loaded db back to file(s). The CLI binaries were mostly added just to support easy unit testing with bash scripts. One thing to also note is that the "compressed" TSVs only keep 48 bits of the MD5 hash.
The songdb code supports adding song lengths etc. to db at runtime, but not the full metadata: https://github.com/mvtiaine/audacious-uade/blob/master/src/songdb/songdb.h#L79

I know about RMC, it's not currently supported though to avoid some extra dependencies (bencodetools, libzakalwe). It would be kind of nice, but then would also need someone to host/maintain/curate the collection. There was some related discussion in EAB also: https://eab.abime.net/showthread.php?t=71058

For the longer term, I have some plans to add a few more possible metadata sources and also to try extract useful metadata from existing sources, but which have it in an unstructured/unparseable form (for example Mods Anthology or even mod file sample texts). Perhaps using an LLM or something would help there :)

**Comment by mvtiaine** [comment]
I did quite a big refactoring/update of the songdb stuff and it now generates pretty printed tsvs also. Also the separate amp/modland/unexotica/tsvs are combined into single metadata tsv (the separate ones are still available as well).
https://github.com/mvtiaine/audacious-uade/pull/37

Next I will split/move the tools directory to a separate repo and add the pretty printed tsvs also there.

Some snippets:

songlengths.tsv
```
fffd7a7d8547	1	250840,p+s
fffd9ea13bc6	1	162860,p
fffdc1d765c3	0	40880,l 117860,l 8780,s 79340,l 8080,s 19000,s
fffdd3c2bef3	1	194960,p
fffe869a7f8d	0	172420,p
fffec16fade0	1	38480,p
```

```
[md5]<TAB>[minsubsong]<TAB>[[songlength(ms),songend]<SPACE>[songlength(ms),songend]<SPACE>[...]]
```

modinfos.tsv
```
fffdc1d765c3	CustomPlay	0
fffdd3c2bef3	Scream Tracker 3.2x (GUS)	8
fffe869a7f8d	AHX v2	0
fffec16fade0	Protracker	0
ffff05daf87b	FastTracker v2.00	12
```

```
[md5]TAB[format]TAB[channels]
```

combined.tsv
```
feaa9d2a4869	Scorpik	Alchemy	Toxic Ziemniak	1992
feaba2f4c992	Jazz			0
feabaabf8a62	Mantronix~Tip	Blue House Productions~Rebels~Sonic Projects	Blue House 2	1991
feabb7b54504	Udo Paschke	APC & TCP	1990	1993
feabc4401195	Basic	Music Contest		1998
```

```
[md5]TAB[authors]TAB[publishers]TAB[album]TAB[year]

multiple authors or publishers are separated by ~
```

**Comment by mvtiaine** [comment]
@devloic I've moved the songdb stuff to https://github.com/mvtiaine/audacious-uade-tools
The pretty printed TSVs are at https://github.com/mvtiaine/audacious-uade-tools/tree/master/tsv/pretty

**Comment by devloic** [comment]
@mvtiaine thanks for the update. I'll try to merge the changes in the fork I just released https://github.com/devloic/audacious-uade. My version is based on 8930ab2b66ae53c9345b01374682387e0da93d64 Thu Jan 2 17:05:16 2025. It includes a binary that basically ports the tsv files to a sqlite database. That makes the data more easily reusable since I needed to use the data in a php app (azuracast).  It is also more flexible so adding album art for example would only require adding a blob field to store the images. 

Did you know about this database: https://modsamplemaster.thegang.nu ? I found it quite interesting.

**Comment by mvtiaine** [comment]
Nice!

I still suspect you would have easier time using the pretty printed TSVs :)
If you want to stick with the c++ code, probably better (to avoid merge issues etc.) is to keep the changes to the songdb.cc code at minimum, e.g. just add methods to dump the internal data, and keep the sqlite export separate.

I saw this https://chipflip.wordpress.com/2019/10/28/mapping-the-world-of-amiga-samples/
but it was more like a prototype back then, looks very cool now.

I've also thought about using some kind of fuzzy hashing, also in the context of finding duplicate or corrupted files (i'd estimate something like 1% of files in various collections are in some way corrupted/bit rotted). It's not unusual to see 3-4 versions of the same mod with exact same sample texts/pattern data/songlength, but still different md5. And then there are cases like original mod from author (maybe even couple of different edits) and multiple ripped versions from some demo by different rippers etc.

**Comment by devloic** [comment]
Yes it would have been a lot easier with the new readable format and the documentation. I missed some data in the process. I feel......a little.....stupid.
You are right the export to sqlite can be kept separate now, no need to reuse songdb.cc code.  Anyway, at the end it will be for something better, just need to find the motivation again.

There are so many amiga music "catalog" sites but no sites actually makes you experience a little bit of the amiga "touch" because well music is about listening, not so much about browsing collections. Anyway I hope I can achieve a little bit of that goal of bringing the amiga alive again. On the long term I'd like to add an emulation part with those wasm uae emulators that can run straight in the browser. The people of emulatorjs ( https://github.com/EmulatorJS/EmulatorJS ) have done a great job so it is all about bringing all those pieces together inside one "experience". I think the machine deserves it.
